### ISSUE TITLE
Track the "rejection" of inline completion suggestions

### ISSUE THREAD

**Issue by spew** [microsoft/vscode#187196]
Several useful changes have been added to the InlineCompletionProvider interface, noteably `handleItemDidShow()` and `handlePartialAccept()`. However, one noteable thing missing is the ability to track if a suggestion was rejected. In the original issue talking about this point, https://github.com/microsoft/vscode/issues/172384, there were two new proposed interface functions, `handleDidRejectCompletionItem()` and `handleDidAcceptCompletionItem()`, can these be added? 

Right now we are indirectly tracking rejections by keeping track of some state and using a combination of providing an acceptance command to `InlineCompletionItem and using node.js's `setTimeout(...)` and assuming that if a suggestion was not accepted after N seconds that it was rejected.

**Comment by spew** [comment]
Wondering if there is an update on this @isidorn -- do you anticipate that `handleItemDidShow()` and `handlePartialAccept()` will be moved out of experimental status any time in the near future?

**Comment by isidorn** [comment]
No updates. We plan to move them out of proposed API, I am just not sure yet. As soon as I learn more I will provide an update here.

Somewhat similar discussion https://github.com/microsoft/vscode/issues/192546

**Comment by spew** [comment]
Thanks, it would be really awesome if vscode could tell us explicitly if a suggestion was rejected, discarded, or accepted. I would consider it 'discarded' if the reason why it was not accepted was that the user continued typing on the same line, etc.

**Comment by Zane-XY** [comment]
Shouldn't the rejection be explicit? Only the user provides explicit rejection action should be considered as a rejected completion. There are implicit discarding or unintentional triggering of the completions, which should not be considered as rejections. 

The hover overlay menu is a good place to provide explicit rejections. I'm not sure if it allows for extension.

**Comment by marrej** [comment]
> Shouldn't the rejection be explicit? Only the user provides explicit rejection action should be considered as a rejected completion. There are implicit discarding or unintentional triggering of the completions, which should not be considered as rejections.

I disagree with this point. If user sees a whole line of completion and they decide to type through it it is explicit rejection.
only if they would not get a chance to see/observe/notice the completion then we could say its not a rejection, for which a plethora of heuristics can be used. E.g. we use a 750ms timer before which we do not treat a suggestion as noticeable.

> The hover overlay menu is a good place to provide explicit rejections. I'm not sure if it allows for extension.

Not really, because not a lot of people even notice/ interact with this menu from my own experience and research. The reject has to be captured as a signal, but can be strengthen by e.g. thumbs up down from the panel to add some additional content/power to the notion.

**Comment by spew** [comment]
I would say hitting ESC is an explicit rejection. Letting a suggestion expire or typing something else does not count as a rejection but counts as discarded.

**Comment by Zane-XY** [comment]
@hediet @isidorn 
When can we expect to have the `handleItemDidShow()` in the InlineCompletionsProvider? Why it's moved out of the proposed API? The times the suggestion show to the user, and the acceptance amount of those shows consists of the important metric to assess the performance the model, currently we lost this important metric.

**Comment by isidorn** [comment]
@Zane-XY no plans in the immediate future. Things in proposed API can be removed, that's what we use proposed API for, so we can try out different things, iterate, and remove not-needed stuff.

**Comment by Zane-XY** [comment]
Hi @isidorn @hediet 
It's not about experimental features, it's something should already provided by the core api. Jetbrains IntelliJ SDK already provided such APIs already, https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/InlineCompletionHandler.kt
In VSCode, we have a way to know if a suggestion is accepted through the command callback, but we don't know how many of the total suggestions are really shown to the user, that's incomplete, we can't compute the acceptance ratio correctly without the `handleItemDidShow()` listener.
